File: infra-market-server/src/main/kotlin/io/infra/market/service/ActivityService.kt
package io.infra.market.service

import io.infra.market.dto.*
import io.infra.market.repository.dao.ActivityDao
import io.infra.market.repository.dao.ActivityTemplateDao
import io.infra.market.repository.entity.Activity
import io.infra.market.repository.entity.ActivityTemplate
import com.fasterxml.jackson.databind.ObjectMapper
import io.infra.structure.core.utils.Loggable
import io.infra.market.util.TimeUtil
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

/**
 * 活动服务
 */
@Service
class ActivityService(
    private val activityDao: ActivityDao,
    private val activityTemplateDao: ActivityTemplateDao,
    private val objectMapper: ObjectMapper
) : Loggable {

    fun findPage(query: ActivityQueryDto): PageResultDto<ActivityDto> {
        val page = activityDao.page(query)
        val activityDtos = page.records.map { convertToDto(it) }
        
        return PageResultDto(
            records = activityDtos,
            total = page.totalRow,
            page = page.pageNumber,
            size = page.pageSize
        )
    }

    fun findById(id: Long): ActivityDto? {
        val activity = activityDao.getById(id)
        return activity?.let { convertToDto(it) }
    }

    @Transactional
    fun save(form: ActivityFormDto): ActivityDto {
        // 验证模板是否存在
        if (form.templateId == null) {
            throw RuntimeException("模板ID不能为空")
        }
        val template = activityTemplateDao.getById(form.templateId)
            ?: throw RuntimeException("活动模板不存在")
        
        // 验证配置数据
        validateConfigData(form.configData, template)
        
        val activity = convertToEntity(form)
        activity.createTime = System.currentTimeMillis()
        activity.updateTime = System.currentTimeMillis()
        activity.status = form.status ?: 1

        activityDao.save(activity)
        return convertToDto(activity)
    }

    @Transactional
    fun update(id: Long, form: ActivityFormDto): ActivityDto {
        val existingActivity = activityDao.getById(id)
            ?: throw RuntimeException("活动不存在")

        // 验证模板是否存在
        if (form.templateId != null) {
            val template = activityTemplateDao.getById(form.templateId)
                ?: throw RuntimeException("活动模板不存在")
            
            // 验证配置数据
            validateConfigData(form.configData, template)
        }

        val activity = convertToEntity(form)
        activity.id = id
        activity.createTime = existingActivity.createTime
        activity.updateTime = System.currentTimeMillis()
        activity.status = form.status ?: existingActivity.status

        activityDao.updateById(activity, false)
        return convertToDto(activity)
    }

    @Transactional
    fun delete(id: Long): Boolean {
        return activityDao.removeById(id)
    }

    @Transactional
    fun updateStatus(id: Long, status: Int): ActivityDto {
        val existingActivity = activityDao.getById(id)
            ?: throw RuntimeException("活动不存在")

        existingActivity.status = status
        existingActivity.updateTime = System.currentTimeMillis()
        activityDao.updateById(existingActivity)
        
        return convertToDto(existingActivity)
    }

    private fun convertToDto(entity: Activity): ActivityDto {
        val configData = parseConfigData(entity.configData)
        
        // 获取模板名称
        var templateName: String? = null
        if (entity.templateId != null) {
            val template = activityTemplateDao.getById(entity.templateId)
            templateName = template?.name
        }

        return ActivityDto(
            id = entity.id,
            name = entity.name,
            description = entity.description,
            templateId = entity.templateId,
            templateName = templateName,
            configData = configData,
            status = entity.status,
            createTime = TimeUtil.format(entity.createTime),
            updateTime = TimeUtil.format(entity.updateTime)
        )
    }

    private fun convertToEntity(form: ActivityFormDto): Activity {
        val configDataJson = try {
            if (form.configData.isNullOrEmpty()) {
                null
            } else {
                objectMapper.writeValueAsString(form.configData)
            }
        } catch (e: Exception) {
            log.error("序列化配置数据失败: ${e.message}", e)
            null
        }

        return Activity(
            name = form.name,
            description = form.description,
            templateId = form.templateId,
            configData = configDataJson,
            status = form.status
        )
    }

    /**
     * 解析配置数据JSON
     * 
     * @param configDataJson 配置数据JSON字符串
     * @return 配置数据Map
     */
    private fun parseConfigData(configDataJson: String?): Map<String, Any>? {
        return try {
            if (configDataJson.isNullOrBlank()) {
                null
            } else {
                @Suppress("UNCHECKED_CAST")
                objectMapper.readValue(configDataJson, Map::class.java) as? Map<String, Any>
            }
        } catch (e: Exception) {
            log.error("解析配置数据失败: ${e.message}", e)
            null
        }
    }

    /**
     * 验证配置数据是否符合模板的字段配置
     * 
     * @param configData 配置数据
     * @param template 活动模板
     */
    private fun validateConfigData(configData: Map<String, Any>?, template: ActivityTemplate) {
        // 解析模板字段配置
        val fields = try {
            if (template.fields.isNullOrBlank()) {
                emptyList()
            } else {
                objectMapper.readValue(template.fields, Array<ActivityTemplateFieldDto>::class.java).toList()
            }
        } catch (e: Exception) {
            log.error("解析模板字段配置失败: ${e.message}", e)
            emptyList()
        }

        // 验证必填字段
        fields.forEach { field ->
            if (field.required == true) {
                val fieldName = field.name ?: return@forEach
                val value = configData?.get(fieldName)
                
                val isEmpty = when (value) {
                    null -> true
                    is String -> value.trim().isEmpty()
                    is Collection<*> -> value.isEmpty()
                    else -> false
                }
                
                if (isEmpty) {
                    val displayName = if (field.label.isNullOrBlank()) {
                        fieldName
                    } else {
                        "${field.label}（${fieldName}）"
                    }
                    throw RuntimeException("字段 $displayName 为必填项，不能为空")
                }
            }
        }
    }
}

File: infra-market-server-dotnet/Services/ActivityService.cs
using System.Text.Json;
using InfraMarket.Server.DTOs;
using InfraMarket.Server.Entities;
using InfraMarket.Server.Repositories.Interfaces;
using InfraMarket.Server.Services.Interfaces;
using InfraMarket.Server.Utils;
using Microsoft.Extensions.Logging;

namespace InfraMarket.Server.Services;

public class ActivityService(
    IActivityRepository activityRepository,
    IActivityTemplateRepository templateRepository,
    ILogger<ActivityService> logger) : IActivityService
{
    public async Task<ApiData<PageResult<ActivityDto>>> GetActivitiesAsync(ActivityQueryDto query)
    {
        try
        {
            var (activities, total) = await activityRepository.PageAsync(query);

            // 批量获取模板信息
            var templateIds = activities.Select(a => a.TemplateId).Distinct().ToList();
            var templateMap = new Dictionary<ulong, ActivityTemplate?>();
            foreach (var templateId in templateIds)
            {
                var template = await templateRepository.FindByIdAsync(templateId);
                templateMap[templateId] = template;
            }

            // 转换为DTO
            var activityDtos = activities.Select(activity =>
            {
                var template = templateMap.GetValueOrDefault(activity.TemplateId);
                var templateName = template?.Name;
                return ConvertActivityToDto(activity, templateName);
            }).ToList();

            var result = new PageResult<ActivityDto>
            {
                Records = activityDtos,
                Total = total,
                Page = query.Page ?? 1,
                Size = query.Size ?? 10
            };

            return ApiData<PageResult<ActivityDto>>.Success(result);
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "获取活动列表失败");
            return ApiData<PageResult<ActivityDto>>.Error("查询失败", 500);
        }
    }

    public async Task<ApiData<ActivityDto>> GetActivityAsync(ulong id)
    {
        try
        {
            var activity = await activityRepository.FindByIdAsync(id);
            if (activity == null)
            {
                logger.LogWarning("获取活动详情失败，活动ID: {ActivityId}，活动不存在", id);
                return ApiData<ActivityDto>.Error("活动不存在", 404);
            }

            // 获取模板信息
            var template = await templateRepository.FindByIdAsync(activity.TemplateId);
            var templateName = template?.Name;

            var activityDto = ConvertActivityToDto(activity, templateName);
            return ApiData<ActivityDto>.Success(activityDto);
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "获取活动详情失败，活动ID: {ActivityId}", id);
            return ApiData<ActivityDto>.Error("查询失败", 500);
        }
    }

    public async Task<ApiData<ActivityDto>> CreateActivityAsync(ActivityFormDto dto)
    {
        try
        {
            // 验证模板是否存在
            var template = await templateRepository.FindByIdAsync(dto.TemplateId);
            if (template == null)
            {
                return ApiData<ActivityDto>.Error("活动模板不存在", 400);
            }

            // 验证配置数据
            var validationError = ValidateConfigData(dto.ConfigData, template);
            if (validationError != null)
            {
                return ApiData<ActivityDto>.Error(validationError, 400);
            }

            // 序列化配置数据
            var configDataJson = SerializeConfigData(dto.ConfigData);

            var activity = new Activity
            {
                Name = dto.Name,
                Description = dto.Description,
                TemplateId = dto.TemplateId,
                ConfigData = configDataJson,
                Status = dto.Status ?? 1
            };

            await activityRepository.CreateAsync(activity);

            var activityDto = ConvertActivityToDto(activity, template.Name);
            return ApiData<ActivityDto>.Success(activityDto);
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "创建活动失败");
            return ApiData<ActivityDto>.Error("创建活动失败", 500);
        }
    }

    public async Task<ApiData<ActivityDto>> UpdateActivityAsync(ulong id, ActivityFormDto dto)
    {
        try
        {
            var activity = await activityRepository.FindByIdAsync(id);
            if (activity == null)
            {
                logger.LogWarning("更新活动失败，活动ID: {ActivityId}，活动不存在", id);
                return ApiData<ActivityDto>.Error("活动不存在", 404);
            }

            // 验证模板是否存在
            if (dto.TemplateId != 0)
            {
                var validationTemplate = await templateRepository.FindByIdAsync(dto.TemplateId);
                if (validationTemplate == null)
                {
                    return ApiData<ActivityDto>.Error("活动模板不存在", 400);
                }

                // 验证配置数据
                var validationError = ValidateConfigData(dto.ConfigData, validationTemplate);
                if (validationError != null)
                {
                    return ApiData<ActivityDto>.Error(validationError, 400);
                }
            }

            // 序列化配置数据
            string? configDataJson = null;
            if (dto.ConfigData != null)
            {
                configDataJson = SerializeConfigData(dto.ConfigData);
            }

            activity.Name = dto.Name;
            if (dto.Description != null)
            {
                activity.Description = dto.Description;
            }
            if (dto.TemplateId != 0)
            {
                activity.TemplateId = dto.TemplateId;
            }
            if (configDataJson != null)
            {
                activity.ConfigData = configDataJson;
            }
            if (dto.Status.HasValue)
            {
                activity.Status = dto.Status.Value;
            }

            await activityRepository.UpdateAsync(activity);

            // 获取模板名称
            var template = await templateRepository.FindByIdAsync(activity.TemplateId);
            var templateName = template?.Name;

            var activityDto = ConvertActivityToDto(activity, templateName);
            return ApiData<ActivityDto>.Success(activityDto);
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "更新活动失败，活动ID: {ActivityId}", id);
            return ApiData<ActivityDto>.Error("更新活动失败", 500);
        }
    }

    public async Task<ApiData<object>> DeleteActivityAsync(ulong id)
    {
        try
        {
            var activity = await activityRepository.FindByIdAsync(id);
            if (activity == null)
            {
                logger.LogWarning("删除活动失败，活动ID: {ActivityId}，活动不存在", id);
                return ApiData<object>.Error("活动不存在", 404);
            }

            await activityRepository.DeleteAsync(id);
            return ApiData<object>.Success(null!);
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "删除活动失败，活动ID: {ActivityId}", id);
            return ApiData<object>.Error("删除活动失败", 500);
        }
    }

    public async Task<ApiData<ActivityDto>> UpdateActivityStatusAsync(ulong id, int status)
    {
        try
        {
            var activity = await activityRepository.FindByIdAsync(id);
            if (activity == null)
            {
                logger.LogWarning("更新活动状态失败，活动ID: {ActivityId}，活动不存在", id);
                return ApiData<ActivityDto>.Error("活动不存在", 404);
            }

            activity.Status = status;
            await activityRepository.UpdateAsync(activity);

            // 获取模板名称
            var template = await templateRepository.FindByIdAsync(activity.TemplateId);
            var templateName = template?.Name;

            var activityDto = ConvertActivityToDto(activity, templateName);
            return ApiData<ActivityDto>.Success(activityDto);
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "更新活动状态失败，活动ID: {ActivityId}，状态: {Status}", id, status);
            return ApiData<ActivityDto>.Error("更新状态失败", 500);
        }
    }

    private static ActivityDto ConvertActivityToDto(Activity activity, string? templateName)
    {
        var configData = ParseConfigData(activity.ConfigData);

        return new ActivityDto
        {
            Id = activity.Id,
            Name = activity.Name,
            Description = activity.Description,
            TemplateId = activity.TemplateId,
            TemplateName = templateName,
            ConfigData = configData,
            Status = activity.Status,
            CreateTime = DateTimeUtil.FormatLocalDateTime(activity.CreateTime),
            UpdateTime = DateTimeUtil.FormatLocalDateTime(activity.UpdateTime)
        };
    }

    private static string? SerializeConfigData(Dictionary<string, object>? configData)
    {
        if (configData == null)
        {
            return null;
        }

        var options = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
            DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull
        };
        return JsonSerializer.Serialize(configData, options);
    }

    private static Dictionary<string, object>? ParseConfigData(string? configDataJson)
    {
        if (string.IsNullOrEmpty(configDataJson))
        {
            return null;
        }

        try
        {
            var options = new JsonSerializerOptions
            {
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
                DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull
            };
            return JsonSerializer.Deserialize<Dictionary<string, object>>(configDataJson, options);
        }
        catch
        {
            return null;
        }
    }

    private static string? ValidateConfigData(Dictionary<string, object>? configData, ActivityTemplate template)
    {
        if (string.IsNullOrEmpty(template.Fields))
        {
            return null;
        }

        try
        {
            var fields = JsonSerializer.Deserialize<List<ActivityTemplateFieldDto>>(template.Fields);
            if (fields == null)
            {
                return null;
            }

            // 验证必填字段
            foreach (var field in fields)
            {
                if (field.Required == true)
                {
                    var fieldName = field.Name;
                    if (string.IsNullOrEmpty(fieldName))
                    {
                        continue;
                    }

                    if (configData == null || !configData.ContainsKey(fieldName))
                    {
                        var displayName = string.IsNullOrEmpty(field.Label) ? fieldName : $"{field.Label}（{fieldName}）";
                        return $"字段 {displayName} 为必填项，不能为空";
                    }

                    var value = configData[fieldName];
                    if (value == null)
                    {
                        var displayName = string.IsNullOrEmpty(field.Label) ? fieldName : $"{field.Label}（{fieldName}）";
                        return $"字段 {displayName} 为必填项，不能为空";
                    }

                    // 检查字符串是否为空
                    if (value is string strValue && string.IsNullOrEmpty(strValue))
                    {
                        var displayName = string.IsNullOrEmpty(field.Label) ? fieldName : $"{field.Label}（{fieldName}）";
                        return $"字段 {displayName} 为必填项，不能为空";
                    }

                    // 检查数组是否为空
                    if (value is System.Collections.ICollection collection && collection.Count == 0)
                    {
                        var displayName = string.IsNullOrEmpty(field.Label) ? fieldName : $"{field.Label}（{fieldName}）";
                        return $"字段 {displayName} 为必填项，不能为空";
                    }
                }
            }

            return null;
        }
        catch
        {
            return null;
        }
    }
}
